File: nnt/lib/core/mapt.dart
part of nnt.core;

class MapT {
  /** 根据查询路径获取值 */
  static dynamic GetValueByKeyPath(Map o, String kp, [dynamic def = null]) {
    if (o == null) {
      return def;
    }
    var ks = kp.split('.');
    var r = any_cast(o);
    for (var i = 0; i < ks.length; ++i) {
      r = r[ks[i]];
      if (r == null) return def;
    }
    return r;
  }

  static dynamic GetValueByKeyPaths(Map o, List<String> ks,
      [dynamic def = null]) {
    if (o == null) {
      return def;
    }
    var r = any_cast(o);
    for (var i = 0; i < ks.length; ++i) {
      r = r[ks[i]];
      if (r == null) return def;
    }
    return r;
  }

  /** 根据查询路径设置值 */
  static bool SetValueByKeyPath(Map o, String kp, dynamic v) {
    if (o == null) {
      print("不能对null进行keypath的设置操作");
      return false;
    }
    final ks = kp.split('.');
    final l = ks.length - 1;
    var r = any_cast(o);
    for (var i = 0; i < l; ++i) {
      var k = ks[i];
      var t = any_cast(r[k]);
      if (t == null) {
        t = {};
        r[k] = t;
      }
      r = t;
    }
    r[ks[l]] = v;
    return true;
  }

  static bool SetValueByKeyPaths(Map o, dynamic v, List<String> ks) {
    if (o == null) {
      print("不能对null进行keypath的设置操作");
      return false;
    }
    final l = ks.length - 1;
    var r = any_cast(o);
    for (var i = 0; i < l; ++i) {
      var k = ks[i];
      var t = any_cast(r[k]);
      if (t == null) {
        t = {};
        r[k] = t;
      }
      r = t;
    }
    r[ks[l]] = v;
    return true;
  }

  // 展开成keypath的结构
  static Map KeyPathExpand(Map o) {
    var r = {};
    _KeyPathExpandAt(o, r, []);
    return r;
  }

  static void _KeyPathExpandAt(dynamic o, Map r, List<String> p) {
    if (IsPod(o)) {
      r[p.join('.')] = o;
      return;
    }

    if (o is List) {
      for (var i = 0, l = o.length; i < l; ++l) {
        var np = List.of(p);
        np.add(i.toString());
        _KeyPathExpandAt(o[i], r, np);
      }
    } else if (o is Map) {
      o.forEach((e, k) {
        var np = List.of(p);
        np.add(k);
        _KeyPathExpandAt(e, r, np);
      });
    } else {
      print('遇到不支持的数据类型');
    }
  }
}

File: nnt-os/android/src/main/kotlin/com/nnt/core/MapT.kt
package com.nnt.core;

class MapT {

    companion object {
        fun Get(o: AnyMap, k: Any, def: Any? = null): Any? {
            if (o.containsKey(k))
                return o[k]
            return def
        }

        /** 根据查询路径获取值 */
        fun GetValueByKeyPath(o: AnyMap?, kp: KeyPath, def: Any? = null): Any? {
            if (o == null) {
                return def;
            }
            val ks = kp.split('.');
            var r = o as Any?
            for (k in ks) {
                if (!(r is MapType)) {
                    return def;
                }
                r = r[k]
                if (r == null) {
                    return def
                }
            }
            return r
        }

        // 获得地一个不为null的key
        fun ValueAtFirstExistsKey(o: AnyMap?, ks: List<String>, def: Any? = null): Any? {
            if (o == null)
                return def
            ks.forEach {
                if (o.containsKey(it))
                    return o[it]
            }
            return def
        }
    }
}

fun toMap(o: Any?, def: Map<*, *>? = null): Map<*, *>? {
    if (o == null)
        return def
    if (o is Map<*, *>) {
        return o
    }
    if (o is String) {
        return toJsonObj(o)
    }
    return def
}